### ISSUE TITLE
[@typescript-eslint/no-loop-func] false positive

### ISSUE THREAD

**Issue by TwitchBronBron** [typescript-eslint/typescript-eslint#2922]
- [x ] I have tried restarting my IDE and the issue persists.
- [x ] I have updated to the latest version of the packages.
- [ x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-loop-func": "error"
  }
}
```

```TS
let text = 'hello';
let words = ['hello', 'world'];
for (let i = 0; i < 10; i++) {
    console.log(
        words.filter(x => x === text)
    );
}
text = 'world';
```

**Expected Result**

There should not be a `@typescript-eslint/no-loop-func` lint error, since the variable is reassigned _after_ the loop and would not actually cause a problem in the loop.

**Actual Result**
There's a `@typescript-eslint/no-loop-func` lint error for `x => x === text` because the variable is reassigned after the loop.

**Additional Info**
![image](https://user-images.githubusercontent.com/2544493/103761784-3af26000-4fe5-11eb-811d-c4bd32e8b38b.png)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.12.0` |
| `@typescript-eslint/parser`        | `4.12.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.17.0` |
| `node`                             | `14.12.0` |

**Comment by bradzacher** [comment]
This is working as expected.
You as an engineer with domain and type knowledge know that `words.filter` is a method which immediately calls the passed callback before it returns.

To a linter however, it just sees a method call. For all it knows `words.filter` could be saving the callback for use later.

Because it doesn't know, it reports.

You can either use a disable comment or update your code to use a `const`.
